pkg/proc/proc.go

```diff
@@ -588,6 +567,53 @@ func GoroutinesInfo(dbp Process, start, count int) ([]*G, int, error) {
 	return allg, -1, nil
 }
 
+func getAllGs(dbp Process) (uint64, uint64, error) {
+	exeimage := dbp.BinInfo().Images[0] // Image corresponding to the executable file
+	rdr := exeimage.DwarfReader()
+
+	addr, err := rdr.AddrFor("runtime.allglen", exeimage.StaticBase)
+	if err != nil {
+		return 0, 0, err
+	}
+	allglenBytes := make([]byte, 8)
+	_, err = dbp.CurrentThread().ReadMemory(allglenBytes, uintptr(addr))
+	if err != nil {
+		return 0, 0, err
+	}
+	allglen := binary.LittleEndian.Uint64(allglenBytes)
+
+	rdr.Seek(0)
+	allgentryaddr, err := rdr.AddrFor("runtime.allgs", exeimage.StaticBase)
+	if err != nil {
+		// try old name (pre Go 1.6)
+		allgentryaddr, err = rdr.AddrFor("runtime.allg", exeimage.StaticBase)
+		if err != nil {
+			return 0, 0, err
+		}
+	}
+	faddr := make([]byte, dbp.BinInfo().Arch.PtrSize())
+	_, err = dbp.CurrentThread().ReadMemory(faddr, uintptr(allgentryaddr))
+	if err != nil {
+		return 0, 0, err
+	}
+	allgptr := binary.LittleEndian.Uint64(faddr)
+
+	return allgptr, allglen, nil
+}
+
+// gorountineAtIndex returns the goroutine at index i in runtime.allgs
+func goroutineAtIndex(p Process, i uint64) (*G, error) {
+	allgptr, allglen, err := getAllGs(p)
+	if i >= allglen {
+		return nil, fmt.Errorf("Index out of bounds")
+	}
+	gvar, err := newGVariable(p.CurrentThread(), uintptr(allgptr+(uint64(i)*uint64(p.BinInfo().Arch.PtrSize()))), true)
```

Comment:
Can we improve the caching by using a *G we already parsed? I understand that the GoID -> index survives longer because it doesn't depend on our internal cache, but is there a possibility to link them somehow? Like if we already have a parsed the goroutine at this index before and it's just sitting in our allgcache for example, can we just figure out a way for the index cache to return that instead of reading the memory all over again? Since we're directly calling `CurrentThread().ReadMemory()` we're not even getting the benefit of memory caching even if we've already read it before.